aws/cloudwatch_dashboard_test.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/aws/aws-sdk-go/service/cloudwatch"
+	"github.com/gruntwork-io/cloud-nuke/config"
+	"github.com/gruntwork-io/cloud-nuke/util"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestListCloudWatchDashboards(t *testing.T) {
+	t.Parallel()
+
+	region, err := getRandomRegion()
+	require.NoError(t, err)
+
+	session, err := session.NewSession(&aws.Config{Region: aws.String(region)})
+	require.NoError(t, err)
+	svc := cloudwatch.New(session)
+
+	cwdbName := createCloudWatchDashboard(t, svc, region)
+	defer deleteCloudWatchDashboard(t, svc, cwdbName, true)
+
+	cwdbNames, err := getAllCloudWatchDashboards(session, time.Now(), config.Config{})
+	require.NoError(t, err)
+	assert.Contains(t, aws.StringValueSlice(cwdbNames), aws.StringValue(cwdbName))
+}
+
+func TestTimeFilterExclusionNewlyCreatedCloudWatchDashboard(t *testing.T) {
+	t.Parallel()
+
+	region, err := getRandomRegion()
+	require.NoError(t, err)
+
+	session, err := session.NewSession(&aws.Config{Region: aws.String(region)})
+	require.NoError(t, err)
+	svc := cloudwatch.New(session)
+
+	cwdbName := createCloudWatchDashboard(t, svc, region)
+	defer deleteCloudWatchDashboard(t, svc, cwdbName, true)
+
+	// Assert CWDB is picked up without filters
+	cwdbNamesNewer, err := getAllCloudWatchDashboards(session, time.Now(), config.Config{})
+	require.NoError(t, err)
+	assert.Contains(t, aws.StringValueSlice(cwdbNamesNewer), aws.StringValue(cwdbName))
+
+	// Assert user doesn't appear when we look at users older than 1 Hour
+	olderThan := time.Now().Add(-1 * time.Hour)
+	cwdbNamesOlder, err := getAllCloudWatchDashboards(session, olderThan, config.Config{})
+	require.NoError(t, err)
+	assert.NotContains(t, aws.StringValueSlice(cwdbNamesOlder), aws.StringValue(cwdbName))
+}
+
+func TestNukeCloudWatchDashboardOne(t *testing.T) {
+	t.Parallel()
+
+	region, err := getRandomRegion()
+	require.NoError(t, err)
+
+	session, err := session.NewSession(&aws.Config{Region: aws.String(region)})
+	require.NoError(t, err)
+	svc := cloudwatch.New(session)
+
+	// We ignore errors in the delete call here, because it is intended to be a stop gap in case there is a bug in nuke.
+	cwdbName := createCloudWatchDashboard(t, svc, region)
+	defer deleteCloudWatchDashboard(t, svc, cwdbName, false)
+	identifiers := []*string{cwdbName}
+
+	require.NoError(
+		t,
+		nukeAllCloudWatchDashboards(session, identifiers),
+	)
+
+	// Make sure the CloudWatch Dashboard is deleted.
+	assertCloudWatchDashboardsDeleted(t, svc, identifiers)
+}
+
+func TestNukeCloudWatchDashboardsMoreThanOne(t *testing.T) {
+	t.Parallel()
+
+	region, err := getRandomRegion()
+	require.NoError(t, err)
+
+	session, err := session.NewSession(&aws.Config{Region: aws.String(region)})
+	require.NoError(t, err)
+	svc := cloudwatch.New(session)
+
+	cwdbNames := []*string{}
+	for i := 0; i < 3; i++ {
+		// We ignore errors in the delete call here, because it is intended to be a stop gap in case there is a bug in nuke.
```

Comment:
Q: for my knowledge - so this stop gap is meant to try & delete the dashboard once, but if that call fails, then it leaves it at that in case the deletion failed for a good reason?

Reply:
Ah no. So the logic here is that the test makes a call to `nukeAllCloudWatchDashboards`, which will go ahead and delete the dashboards that it created in the test, if it works. However, if there is a bug in the function and it fails, those dashboards will be lying around. To avoid having the test pollute the environment with those dashboards, we attempt an alternative cleaning function here to try to make sure those get deleted.

The reason we can't error check is that in the happy path, the dashboards would be deleted by the nuke function and they won't exist, thus this function will throw an error.

Reply:
ahhh! that makes total sense! thanks, Yori :)